cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java

```diff
@@ -494,6 +500,124 @@ public void testBranchLoopLineage() throws Exception {
+    //noinspection ToArrayCallWithZeroLengthArrayArgument
+    metadataStore.addTags(MetadataScope.USER, dataset1, run1Data1Meta.getTags().toArray(new String[0]));
+    metadataStore.setProperties(MetadataScope.USER, dataset2, run1Data2Meta.getProperties());
+    //noinspection ToArrayCallWithZeroLengthArrayArgument
+    metadataStore.addTags(MetadataScope.USER, dataset2, run1Data2Meta.getTags().toArray(new String[0]));
+    TimeUnit.MILLISECONDS.sleep(1);
+
+    // Add accesses for D3 -> P2 -> D2 -> P1 -> D1 <-> P3
+    // We need to use current time here as metadata store stores access time using current time
+    ProgramRunId run1 = program1.run(RunIds.generate(System.currentTimeMillis()).getId());
+    ProgramRunId run2 = program2.run(RunIds.generate(System.currentTimeMillis()).getId());
+    ProgramRunId run3 = program3.run(RunIds.generate(System.currentTimeMillis()).getId());
+
+    ProgramRunId workflow = program6.run(RunIds.generate(System.currentTimeMillis()).getId());
+
+    ProgramRunId run5 = program5.run(RunIds.generate(System.currentTimeMillis()).getId());
+
+    addWorkflowRuns(store, workflow.getProgram(), workflow.getRun(), run1, run2, run3);
+    addRuns(store, workflow);
+    addRuns(store, run5);
+
+    // It is okay to use current time here since access time is ignore during assertions
+    lineageStore.addAccess(run1, dataset1, AccessType.WRITE, System.currentTimeMillis(), flowlet1);
+    lineageStore.addAccess(run1, dataset1, AccessType.WRITE, System.currentTimeMillis(), flowlet1);
+    lineageStore.addAccess(run1, dataset2, AccessType.READ, System.currentTimeMillis(), flowlet1);
+
+    lineageStore.addAccess(run2, dataset2, AccessType.WRITE, System.currentTimeMillis(), flowlet2);
+    lineageStore.addAccess(run2, dataset3, AccessType.READ, System.currentTimeMillis(), flowlet2);
+
+    lineageStore.addAccess(run3, dataset1, AccessType.UNKNOWN, System.currentTimeMillis());
+
+    lineageStore.addAccess(run5, dataset1, AccessType.READ, System.currentTimeMillis());
+
+
+    // The UNKNOWN access type will get filtered out if there is READ/WRITE. It will be preserved if it is the
+    // only access type
+    Lineage expectedLineage = new Lineage(
+      ImmutableSet.of(
+        new Relation(dataset1, program6, AccessType.WRITE, twillRunId(workflow)),
+        new Relation(dataset2, program6, AccessType.READ, twillRunId(workflow)),
+        new Relation(dataset2, program6, AccessType.WRITE, twillRunId(workflow)),
+        new Relation(dataset3, program6, AccessType.READ, twillRunId(workflow)),
+        new Relation(dataset1, program6, AccessType.UNKNOWN, twillRunId(workflow)),
+        new Relation(dataset1, program5, AccessType.READ, twillRunId(run5))
+      )
+    );
+
+    Lineage resultLineage = lineageAdmin.computeLineage(dataset1, 500, System.currentTimeMillis() + 10000,
+                                                        100, "workflow");
```

Comment:
The other tests like the testSimpleLineage etc. should be testing for those use cases, right ?